base_layer/core/src/transactions/transaction_protocol/sender.rs

```diff
@@ -702,12 +703,12 @@ impl fmt::Display for SenderTransactionProtocol {
 }
 
 pub fn calculate_tx_id<D: Digest>(pub_nonce: &PublicKey, index: usize) -> TxId {
```

Comment:
We have a slight problem here as `pub fn calculate_tx_id<D: Digest>` is called from `pub fn build<D: Digest>` which in turn is called from many places where the digest type is specified by the client. If the digest were to be changed where `build` is called this will have no effect.

Reply:
One way to fix this is to simply remove the digest type `<D: Digest>` from `build` and from `calculate_tx_id` because `CalculateTxIdTransactionProtocolHasherBlake256` is already fixed to digest type `Blake256`. This will enhance the robustness of the code as `TxId` should only be hashed in one way.

Reply:
If the generic digest type is a requirement (which I cannot see why at the moment) then we could make use of the `DomainDigest` type as 
``` rust
pub fn calculate_tx_id<D: Digest + DomainDigest>(pub_nonce: &PublicKey, index: usize) -> TxId {
    let hash = D::new()
        .chain(pub_nonce.as_bytes())
        .chain(index.to_le_bytes())
        .finalize();
    let mut bytes: [u8; 8] = [0u8; 8];
    bytes.copy_from_slice(&hash.as_ref()[..8]);
    u64::from_le_bytes(bytes).into()
}
```
and call the function as 
``` rust
calculate_tx_id::<TxIdHasher<D>>(...)
```
with 
``` rust
pub type TxIdHasher<D> = DomainSeparatedHasher<D, TxIdHashDomain>;
```

Reply:
Very good observation. I made changes following the first approach, just delete the generic parameter for now, as it seems to be sufficient for our main purposes. Although I took note of the second approach as it might be useful in the future.